doozer/doozerlib/cli/scan_sources.py

```diff
@@ -54,6 +68,143 @@ def run(self):
+                self.changing_rpm_metas.add(metadata)
+
+            elif metadata.meta_type == 'image':
+                self.add_image_meta_change(metadata, rebuild_hint)
+
+            else:
+                raise IOError(f'Unsupported meta type: {metadata.meta_type}')
+
+        except ChildProcessError:
+            # Failed pushing to openshift-priv
+            self.runtime.logger.warning('failed pushing to openshift-priv for %s', metadata.name)
+            self.issues.append({'name': metadata.distgit_key,
+                                'issue': 'Failed pushing to openshift-priv'})
+
+    def _do_shas_match(self, public_org, priv_org, repo_name, branch_name) -> bool:
+        """
+        Use GitHub API to check commit SHAs on private and public upstream for a given branch.
+        Return True if they match, False otherwise
+        """
+
+        # Check public commit ID
+        pub_commit = self.github_client.get_repo(f'{public_org}/{repo_name}').get_branch(branch_name).commit
+
+        # Check private commit ID
+        priv_commit = self.github_client.get_repo(f'{priv_org}/{repo_name}').get_branch(branch_name).commit
+
+        if pub_commit.sha == priv_commit.sha:
+            self.runtime.logger.debug('Latest commits match on public and priv upstreams for %s', repo_name)
+            return True
+
+        self.runtime.logger.info('Latest commits do not match on public and priv upstreams for %s: '
+                                 'public SHA = %s, private SHA = %s', repo_name, pub_commit.sha, priv_commit.sha)
+        return False
+
+    def _is_pub_ancestor_of_priv(self, path: str, branch_name: str, repo_name: str) -> bool:
+        """
+        If a reconciliation already happened, private upstream might have a merge commit thus be a descendant
+        of the public upstream. In this case, we don't need to rebase public into priv
+
+        Use merge-base --is-ancestor to determine if public upstream is an ancestor of the private one
+        """
+
+        with Dir(path):
+            rc, _, _ = exectools.cmd_gather(['git', 'merge-base', '--is-ancestor',
+                                             f'public_upstream/{branch_name}', f'origin/{branch_name}'])
+            if rc:
+                self.runtime.logger.info('Public upstream is ahead of private for %s: will need to rebase', repo_name)
+                return False
+
+        self.runtime.logger.info('Private upstream is ahead of public for %s: no need to rebase', repo_name)
+        return True
+
+    def rebase_into_priv(self):
+        self.runtime.logger.info('Rebasing public upstream contents into openshift-priv')
+
+        upstream_mappings = exectools.parallel_exec(
+            lambda meta, _: (meta, self.runtime.get_public_upstream(meta.config.content.source.git.url)),
+            self.all_metas,
+            n_threads=20,
+        ).get()
+
+        for metadata, public_upstream in upstream_mappings:
+            if not public_upstream:
```

Comment:
Done with 5afb92d653e7c08ccee12bfd2a98fd107b5e2025

I checked that, like you said, in case `public_upstreams` is not  defined that function will just return the same upstreams as defined in `content.source.git.url`. In this case, we can skip the rebase if the public upstream (as returned by `get_public_upstream()`) match the private one